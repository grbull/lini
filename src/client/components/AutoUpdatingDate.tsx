import React, { ReactElement, useState } from 'react';

import { useInterval } from '../hooks/useInterval';

interface Props {
  className?: string;
  date: string | null;
  format: (date: string) => string;
}

export function AutoUpdatingDate({
  className,
  date,
  format,
}: Props): ReactElement {
  const [formattedDate, setFormattedDate] = useState(
    date ? format(date) : 'N/A'
  );

  useInterval(() => {
    if (date) {
      setFormattedDate(format(date));
    }
    // Updates date every 10min.
  }, 1000 * 60 * 10);

  return (
    <time className={className} dateTime={date || undefined}>
      {formattedDate}
    </time>
  );
}

AutoUpdatingDate.defaultProps = { className: undefined };
