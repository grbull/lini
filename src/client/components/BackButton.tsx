import { faArrowLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import cn from 'classnames';
import React, { ReactElement } from 'react';
import { useHistory } from 'react-router';

interface Props {
  className?: string;
}

export function BackButton({ className }: Props): ReactElement {
  const history = useHistory();

  function handleClick(): void {
    history.goBack();
  }

  return (
    <button
      className={cn('w-11', 'float-left', className)}
      onClick={handleClick}
      tabIndex={0}
      type="button"
    >
      <FontAwesomeIcon className={cn('m-1', 'text-lg')} icon={faArrowLeft} />
    </button>
  );
}

BackButton.defaultProps = { className: undefined };
