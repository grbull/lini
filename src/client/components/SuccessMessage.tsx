import cn from 'classnames';
import React, { ReactElement, ReactNode } from 'react';

interface Props {
  children: string | ReactNode;
  className?: string;
  illustration?: boolean;
}

export function SuccessMessage({
  children,
  className,
  illustration,
}: Props): ReactElement {
  return (
    <div
      className={cn(
        'px-2',
        'py-2.5',
        'bg-green-400',
        'bg-opacity-50',
        'border',
        'border-solid',
        'border-green-400',
        className
      )}
    >
      <div
        className={cn('flex', 'flex-col', 'items-center', 'justify-center', {
          hidden: !illustration,
        })}
      >
        <img
          alt="Happy Illustration"
          className={cn('opacity-80', 'mb-2.5')}
          src="happy.png"
          width={80}
        />
      </div>
      <h2 className={cn('font-bold')}>Success</h2>
      <p>{children}</p>
    </div>
  );
}

SuccessMessage.defaultProps = { className: undefined, illustration: false };
