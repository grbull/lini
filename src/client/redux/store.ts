import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { episode } from './episode';
import { pushSubscription } from './push_subscription';
import { schedule } from './schedule';
import { search } from './search';
import { show } from './show';
import { subscription } from './subscription';
import { user } from './user';

export const reducers = combineReducers({
  user: user.reducer,
  search: search.reducer,
  subscription: subscription.reducer,
  show: show.reducer,
  episode: episode.reducer,
  schedule: schedule.reducer,
  pushSubscription: pushSubscription.reducer,
});

const persistConfig = {
  key: 'root',
  storage,
  version: 2,
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
  reducer: persistedReducer,
  // https://github.com/rt2zz/redux-persist/issues/988#issuecomment-552242978
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  devTools: process.env.NODE_ENV === 'development',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
