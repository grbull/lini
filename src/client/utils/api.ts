import Axios from 'axios';

import { AuthLoginDto, AuthValidateDto } from '../../server/auth/auth.dto';
import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { ShowDto, ShowWithEpisodesDto } from '../../server/show/show.dto';
import {
  SubscriptionCreateDto,
  SubscriptionDto,
  SubscriptionRemoveDto,
  // SubscriptionUpdateDto,
} from '../../server/subscription/subscription.dto';
import { UserDto, UserUpdateDto } from '../../server/user/user.dto';

const axios = Axios.create({
  baseURL: process.env.SERVER_URL,
  withCredentials: true,
});

async function userLogin(loginDto: AuthLoginDto): Promise<void> {
  await axios.post<void>('auth/login', loginDto);
}

async function userValidate(validateDto: AuthValidateDto): Promise<UserDto> {
  const response = await axios.post<UserDto>('auth/validate', validateDto);
  return response.data;
}

async function userGet(): Promise<UserDto> {
  const response = await axios.get<UserDto>('user');
  return response.data;
}

async function userUpdate(updateDto: UserUpdateDto): Promise<UserDto> {
  const response = await axios.put<UserDto>('user', updateDto);
  return response.data;
}

async function showSearch(query: string): Promise<ShowDto[]> {
  const response = await axios.get<ShowDto[]>(`show/search/${query}`);
  return response.data;
}

async function showGet(id: number): Promise<ShowWithEpisodesDto> {
  const response = await axios.get<ShowWithEpisodesDto>(`show/${id}`);
  return response.data;
}

async function episodeGet(id: number): Promise<EpisodeDto> {
  const response = await axios.get<EpisodeDto>(`episode/${id}`);
  return response.data;
}

async function episodeGetByShow(id: number): Promise<EpisodeDto[]> {
  const response = await axios.get<EpisodeDto[]>(`episode/show/${id}`);
  return response.data;
}

async function subscriptionGetAll(): Promise<SubscriptionDto[]> {
  const response = await axios.get<SubscriptionDto[]>('subscription/all');
  return response.data;
}

async function subscriptionCreate(
  createDto: SubscriptionCreateDto
): Promise<SubscriptionDto> {
  const response = await axios.post<SubscriptionDto>('subscription', createDto);
  return response.data;
}

// async function subscriptionUpdate(
//   updateDto: SubscriptionUpdateDto
// ): Promise<SubscriptionDto> {
//   const response = await axios.put<SubscriptionDto>('subscription', updateDto);
//   return response.data;
// }

async function subscriptionRemove(
  removeDto: SubscriptionRemoveDto
): Promise<SubscriptionDto> {
  const response = await axios.delete<SubscriptionDto>('subscription', {
    data: removeDto,
  });
  return response.data;
}

async function episodeScheduleGet(): Promise<EpisodeScheduleDto> {
  const response = await axios.get<EpisodeScheduleDto>('episode/schedule');
  return response.data;
}

async function pushCreate(createDto: PushSubscriptionCreateDto): Promise<void> {
  await axios.post('push_subscription', createDto);
}

export const api = {
  user: {
    login: userLogin,
    validate: userValidate,
    get: userGet,
    update: userUpdate,
  },
  show: {
    search: showSearch,
    get: showGet,
  },
  episode: {
    get: episodeGet,
    getByShow: episodeGetByShow,
    getSchedule: episodeScheduleGet,
  },
  subscription: {
    getAll: subscriptionGetAll,
    create: subscriptionCreate,
    // update: subscriptionUpdate,
    remove: subscriptionRemove,
  },
  push: {
    create: pushCreate,
  },
};
