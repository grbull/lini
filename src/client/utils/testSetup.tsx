import { configureStore } from '@reduxjs/toolkit';
import { render, RenderResult } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { reducers, RootState } from '../redux/store';

export interface TestSetupProps {
  history?: MemoryHistory;
  state?: Partial<RootState>;
}

export interface TestSetupResult extends RenderResult {
  dispatch: typeof jest.fn;
  history: MemoryHistory<unknown>;
  scrollTo: typeof jest.fn;
}

export function testSetup(
  children: ReactNode,
  props?: TestSetupProps
): TestSetupResult {
  const history = props?.history || createMemoryHistory();
  const store = configureStore({
    reducer: reducers,
    preloadedState: props?.state || undefined,
  });

  // const reduxHistory = [];
  // const dispatch = (action: any): any => reduxHistory.push(action);
  const dispatch = jest.fn();
  store.dispatch = dispatch;

  const scrollTo = jest.fn();
  window.scrollTo = scrollTo;

  const utils = render(
    <Router history={history}>
      <Provider store={store}>{children}</Provider>
    </Router>
  );

  return { dispatch, history, scrollTo, ...utils };
}
