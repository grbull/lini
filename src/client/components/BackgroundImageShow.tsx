import cn from 'classnames';
import React, { ReactElement, ReactNode } from 'react';

import { useIsOnline } from '../hooks/useIsOnline';
import { imageShow } from '../utils/imageShow';

interface Props {
  children: ReactNode;
  imageURL: string | null;
}

export function BackgroundImageShow({
  children,
  imageURL,
}: Props): ReactElement {
  const isOnline = useIsOnline();
  return (
    <div
      className={cn(
        'absolute',
        'bg-scroll',
        'bg-cover',
        'bg-center',
        'bg-scroll',
        'w-full md:max-w-4xl',
        'mx-0 md:mx-auto',
        'h-full',
        'left-0 md:left-auto',
        'top-0',
        'pt-12 md:pt-14',
        { 'pb-14 md:pb-8': isOnline },
        { 'pb-20 md:pb-14': !isOnline }
      )}
      style={{ backgroundImage: `url(${imageShow(imageURL)})` }}
    >
      <div
        className={cn(
          'w-full',
          'h-full',
          'bg-gray-100 dark:bg-gray-700',
          'bg-opacity-30 dark:bg-opacity-80',
          'overflow-y-auto'
        )}
      >
        {children}
      </div>
    </div>
  );
}
