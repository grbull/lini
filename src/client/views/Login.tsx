import cn from 'classnames';
import React, { ReactElement, useState } from 'react';
import { Helmet } from 'react-helmet';
import { useSelector } from 'react-redux';
import { Redirect } from 'react-router';

import { ErrorMessage } from '../components/ErrorMessage';
import { LoginForm } from '../components/LoginForm';
import { RootState } from '../redux/store';
import { api } from '../utils/api';

export function Login(): ReactElement {
  const [isSuccess, setIsSuccess] = useState<boolean | undefined>(undefined);

  // Need to display loading some how

  function submitHandler(email: string): void {
    api.user
      .login({ email })
      .then(() => setIsSuccess(true))
      .catch(() => setIsSuccess(false));
  }

  const user = useSelector((state: RootState) => state.user);
  if (user.data && !user.error) {
    return <Redirect to="/" />;
  }

  return (
    <>
      <Helmet>
        <title>Login - Lini</title>
      </Helmet>

      <main
        className={cn(
          'h-screen',
          'flex',
          'flex-col',
          'justify-center',
          'items-center',
          'px-8'
        )}
      >
        <div className={cn('h-64')}>
          <h1 className={cn('text-2xl', 'mb-6', 'text-center')}>Login</h1>
          {isSuccess === undefined && <LoginForm onSubmit={submitHandler} />}
          {isSuccess === true && <div>Success, please check your email</div>}
          {isSuccess === false && (
            <ErrorMessage className={cn('mb-3')}>
              A link was recently issued.
            </ErrorMessage>
          )}
        </div>
      </main>
    </>
  );
}
