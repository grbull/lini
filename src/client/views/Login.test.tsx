/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { fireEvent, waitFor } from '@testing-library/react';
import React from 'react';
import { Route } from 'react-router-dom';

import { SuccessMessage } from '../components/SuccessMessage';
import { RootState } from '../redux/store';
import { api } from '../utils/api';
import { testSetup } from '../utils/testSetup';
import { Login } from './Login';

describe('Login View', () => {
  it('matches the snapshot', () => {
    const { asFragment } = testSetup(<Login />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot when logged in', () => {
    const initialState: Partial<RootState> = {
      user: {
        status: 'idle',
        isLoggedIn: true,
        data: {
          email: 'test@email.com',
          theme: 'auto',
          notifications: false,
          dateCreated: '',
          dateModified: '',
        },
      },
    };

    const { asFragment } = testSetup(
      <>
        <Route component={Login} />
        <Route component={SuccessMessage} exact path="/" />
      </>,
      { state: initialState }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot when api call is success', async () => {
    const mockedFetch = jest
      .spyOn(api.user, 'login')
      .mockImplementationOnce(() => Promise.resolve());

    const { asFragment, getByLabelText, getByText } = testSetup(<Login />);
    const input = getByLabelText('email-input') as HTMLInputElement;
    const button = getByText('Request Link') as HTMLButtonElement;

    fireEvent.change(input, { target: { value: 'test@email.com' } });

    await waitFor(() => expect(button).toBeEnabled());

    fireEvent.click(button);

    await waitFor(() => expect(mockedFetch).toBeCalled());

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot when api call is error', async () => {
    const mockedFetch = jest
      .spyOn(api.user, 'login')
      .mockImplementationOnce(() => Promise.reject());

    const { asFragment, getByLabelText, getByText } = testSetup(<Login />);
    const input = getByLabelText('email-input') as HTMLInputElement;
    const button = getByText('Request Link') as HTMLButtonElement;

    fireEvent.change(input, { target: { value: 'test@email.com' } });

    await waitFor(() => expect(button).toBeEnabled());

    fireEvent.click(button);

    await waitFor(() => expect(mockedFetch).toBeCalled());

    expect(asFragment()).toMatchSnapshot();
  });
});
