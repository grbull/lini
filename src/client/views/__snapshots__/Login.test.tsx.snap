// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login View matches the snapshot 1`] = `
<DocumentFragment>
  <main
    class="absolute left-1/2 top-1/2 m-auto w-80 h-96 -ml-40 -mt-48"
  >
    <img
      alt="logo"
      class="w-28 mx-auto"
      src="/android-chrome-192x192.png"
    />
    <h1
      class="text-2xl my-6 text-center"
    >
      Login
    </h1>
    <form>
      <p
        class="mb-6"
      >
        Enter your email address and you will be sent a link to login with.
      </p>
      <input
        aria-label="email-input"
        autocomplete="email"
        class="w-full mb-6 p-2 text-gray-600 bg-gray-200 border-2 border-solid border-gray-200 hover:border-blue-300 focus:border-blue-300"
        name="email"
        placeholder="Email"
        value=""
      />
      <button
        class="w-28 py-2 border-2 border-solid border-green-200 text-green-200 hover:border-green-300 hover:text-green-300 disabled:opacity-50 disabled:border-gray-200 disabled:text-gray-200 disabled:cursor-not-allowed"
        disabled=""
        type="button"
      >
        Request Link
      </button>
    </form>
  </main>
</DocumentFragment>
`;

exports[`Login View matches the snapshot when api call is error 1`] = `
<DocumentFragment>
  <main
    class="absolute left-1/2 top-1/2 m-auto w-80 h-96 -ml-40 -mt-48"
  >
    <img
      alt="logo"
      class="w-28 mx-auto"
      src="/android-chrome-192x192.png"
    />
    <h1
      class="text-2xl my-6 text-center"
    >
      Login
    </h1>
    <div
      class="px-2 py-2.5 bg-red-400 bg-opacity-50 border border-solid border-red-400 mb-3"
    >
      <h2
        class="font-bold"
      >
        Error
      </h2>
      <p>
        A link was recently issued.
      </p>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Login View matches the snapshot when api call is success 1`] = `
<DocumentFragment>
  <main
    class="absolute left-1/2 top-1/2 m-auto w-80 h-96 -ml-40 -mt-48"
  >
    <img
      alt="logo"
      class="w-28 mx-auto"
      src="/android-chrome-192x192.png"
    />
    <h1
      class="text-2xl my-6 text-center"
    >
      Login
    </h1>
    <div
      class="px-2 py-2.5 bg-green-400 bg-opacity-50 border border-solid border-green-400 mb-3"
    >
      <div
        class="flex flex-col items-center justify-center hidden"
      >
        <img
          alt="Happy Illustration"
          class="opacity-80 mb-2.5"
          src="happy.png"
          width="80"
        />
      </div>
      <h2
        class="font-bold"
      >
        Success
      </h2>
      <p>
        Success, please check your email.
      </p>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Login View matches the snapshot when logged in 1`] = `
<DocumentFragment>
  <div
    class="px-2 py-2.5 bg-green-400 bg-opacity-50 border border-solid border-green-400"
  >
    <div
      class="flex flex-col items-center justify-center hidden"
    >
      <img
        alt="Happy Illustration"
        class="opacity-80 mb-2.5"
        src="happy.png"
        width="80"
      />
    </div>
    <h2
      class="font-bold"
    >
      Success
    </h2>
    <p />
  </div>
</DocumentFragment>
`;
